ethers-providers/src/rpc/transports/ws/manager.rs

```diff
@@ -0,0 +1,373 @@
+use std::{
+    collections::{BTreeMap, HashMap},
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc, Mutex,
+    },
+};
+
+use ethers_core::types::U256;
+use futures_channel::{mpsc, oneshot};
+use futures_util::{select_biased, StreamExt};
+use serde_json::value::RawValue;
+
+use crate::JsonRpcError;
+
+use super::{
+    backend::{BackendDriver, WsBackend},
+    ActiveSub, ConnectionDetails, InFlight, Instruction, Notification, PubSubItem, Response, SubId,
+    WsClient, WsClientError,
+};
+
+pub type SharedChannelMap = Arc<Mutex<HashMap<U256, mpsc::UnboundedReceiver<Box<RawValue>>>>>;
+
+pub const DEFAULT_RECONNECTS: usize = 5;
+
+pub struct SubscriptionManager {
+    subs: BTreeMap<u64, ActiveSub>,
+    aliases: HashMap<U256, u64>,
```

Comment:
realize it's a chore but do you mind adding docs from the PR description on the motivation on why we need to wire together the internal representation of subscriptions to the IDs we get on each reconnect? OK to make it a follow-up or open an issue

Reply:
ack. I am adding all commentary from the pr description to the relevant structs.